Bugfix/Glucose data popup on chart cut off at top of screen (in Trio dev branch)

taylorpatterson-T1D opened the pull request:
Bug Root Cause: The SelectionPopoverView had:

.overflowResolution with y-axis set to .disabled, preventing automatic repositioning
Larger fonts and padding that made the popup too big
Solution Applied:
Fixed Positioning Logic (SelectionPopoverView.swift:41):
- Changed overflowResolution: .init(x: .fit(to: .chart), y: .disabled)
- To overflowResolution: .init(x: .fit(to: .chart), y: .fit(to: .chart))
- This enables automatic repositioning when the popup would go off-screen
Made Popup More Compact:
- Reduced font sizes from .body to .caption and .caption2
- Decreased padding from .horizontal (16px) to .horizontal(8)
- Reduced vertical padding from 2px to 4px total
- Reduced spacing between elements
- Made icon frames smaller (15px → 12px)
- Reduced stroke width from 2 to 1.5

Additional fix for Stats > Meals - chart's legend was wrapping text. Fixed and tested also on smaller screen.

taylorpatterson-T1D committed:
Merge pull request #1 from TaylorJPatterson/dev

Push dev changes into my fork of main

taylorpatterson-T1D committed:
Bugfix/Glucose data popup on chart cut off at top of screen.

Root Cause: The SelectionPopoverView had:

.overflowResolution with y-axis set to .disabled, preventing automatic repositioning
Larger fonts and padding that made the popup too big
Solution Applied:

Fixed Positioning Logic (SelectionPopoverView.swift:41):
- Changed overflowResolution: .init(x: .fit(to: .chart), y: .disabled)
- To overflowResolution: .init(x: .fit(to: .chart), y: .fit(to: .chart))
- This enables automatic repositioning when the popup would go off-screen
Made Popup More Compact:
- Reduced font sizes from .body to .caption and .caption2
- Decreased padding from .horizontal (16px) to .horizontal(8)
- Reduced vertical padding from 2px to 4px total
- Reduced spacing between elements
- Made icon frames smaller (15px → 12px)
- Reduced stroke width from 2 to 1.5

taylorpatterson-T1D committed:
Update Localizable.xcstrings

taylorpatterson-T1D committed:
Update Localizable.xcstrings

taylorpatterson-T1D commented:
@dnzxy I think I correctly rebased in order to submit just these 4 files that repair the glucose popover in the Trio dev branch. Please tell me if this will work or not.

taylorpatterson-T1D committed:
feat/AutoApplyOverride for Trio

Key Capabilities:

Motion-based activity detection (walking, running, cycling) so Trio will automatically apply personal overrides in case user forgets.

Configurable timing (minimum duration before activation, stop duration)
Activity logging with history
Full integration with Trio's settings and override systems
Proper permission handling for CoreMotion

Components Added:
Settings Model (TrioSettings.swift):
- Added ActivityType enum with walking/running/cycling/other
- Added ActivityLogEntry struct for tracking activities
- Added auto-apply settings fields (enable flags, override mappings, timing settings)

Activity Detection Engine (ActivityDetectionManager.swift):
- CoreMotion integration for activity detection
- Delegate pattern for callbacks
- Timer-based confirmation and stop logic
- Permission handling and error management
- Activity logging with persistence

Settings Module (AutoApplyOverride/):
- Complete modular implementation following Trio's architecture
- Override preset selection per activity type
- Timing controls and activity log display

Integration (APSManager.swift):
- Added ActivityDetectionDelegate implementation
- Automatic override application/cancellation
- Integration with existing override system
- Proper error handling

Navigation (Screen.swift, FeatureSettingsView.swift):
- Added navigation routing
- Settings menu integration under "Trio Features"

Architecture Compliance:
All changes contained within /Trio submodule
Follows Trio's modular pattern
Uses existing override system and storage
Integrates with dependency injection
Maintains JSON-based settings persistence

Key Error Handling Applied:
Thread Safety for Timers: Timer scheduling is in main thread to prevent
thread-related crashes
Core Data Threading: Uses background context instead of main context for Core Data
operations to prevent UI blocking
Activity Log Thread Safety: Included proper dispatch queues for all activity log
operations
Error Handling: Added proper error handling for file storage operations
Retain Cycle Prevention: Enhanced weak self references in async operations

Sjoerd-Bo3 commented:
Yes, this is better. But the changes you are proposing in the changes are not aligned with the description you are posting. It seems you are also introducing a feature to activate overrides upon detecting activity. Which is cool but not an fix to the issue you are proposing here.

The way to do this: Make a PR for every fix and Feature seperate. So its clear what we need to look at and test. When multiple features or fixes are in a PR it’s a pain to test/review. Also this is the way its generally done is Software development.

Can you do that? Or do you need some help with it? I’m up to help you, just contact me on discord to converse quickly.

dnzxy commented on Trio/Resources/Info.plist:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/APS/APSManager.swift:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/Assemblies/ServiceAssembly.swift:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/Assemblies/UIAssembly.swift:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/Models/TrioSettings.swift:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

Additionally, the enum and struct do not belong in here, but should be file(s) of their own.

dnzxy commented on Trio/Sources/Modules/AutoApplyOverride/AutoApplyOverrideDataFlow.swift:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/Modules/AutoApplyOverride/AutoApplyOverrideProvider.swift:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/Modules/AutoApplyOverride/AutoApplyOverrideStateModel.swift:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/Modules/AutoApplyOverride/View/AutoApplyOverrideRootView.swift:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/Modules/Home/View/Chart/ChartElements/SelectionPopoverView.swift:
```diff
@@ -61,50 +61,64 @@ struct SelectionPopoverView: ChartContent {
-                .foregroundStyle(Color.insulin).font(.body)
+                .foregroundStyle(Color.insulin)
+                .font(.caption2)
+                .transition(.opacity.combined(with: .scale(scale: 0.95)))
```
Why are you making this (partially) opaque?

taylorpatterson-T1D replied:
so that when sliding over the x axis scale on the right side you can still see the numbers if the popover covers it.
![partially opaque](https://github.com/user-attachments/assets/27ee24df-850c-474d-9460-91faa4b738a3)

Sjoerd-Bo3 replied:
If your changing things like these:
- add a before and after pics in the PR description of everything you changed and how it looks.
- add the change in the PR description (there is no mention of the opaque change in there)

You can also modify something like this in a FIX PR, but maybe just call it differently all together so it reflects what you are doing and where.

dnzxy commented on Trio/Sources/Modules/Home/View/Chart/ChartElements/SelectionPopoverView.swift:
```diff
@@ -61,50 +61,64 @@ struct SelectionPopoverView: ChartContent {
-                .foregroundStyle(Color.orange).font(.body)
+                .foregroundStyle(Color.orange)
+                .font(.caption2)
+                .transition(.opacity.combined(with: .scale(scale: 0.95)))
```
Same question here.

dnzxy commented on Trio/Sources/Modules/Home/View/Chart/MainChartView.swift:
Not sure I agree with these refactorings here. They aren't really necessary and not related to the issue you are proposing to fix. Could you provide some more context here, as to why you are making these changes?

taylorpatterson-T1D replied:
AI made this adaptation as a result of me asking it to make the transition between data points faster, they were lagging like an old school alarm clock (flip counter odometer effect). This was AI's attempt to make it more responsive as the user swipes along the graph.

dnzxy commented on Trio/Sources/Modules/Settings/View/SettingsRootView.swift:
Unnecessary linting change, please revert.

dnzxy commented on Trio/Sources/Modules/Settings/View/Subviews/FeatureSettingsView.swift:
Unnecessary change, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/Modules/Stat/View/ViewElements/Meal/MealStatsView.swift:
```diff
@@ -215,7 +215,7 @@ struct MealStatsView: View {
             ] : [(String(localized: "Carbs"), Color.orange)]
 
-            let columns = [GridItem(.adaptive(minimum: 65), spacing: 4)]
+            let columns = [GridItem(.adaptive(minimum: 90), spacing: 4)]
```
Why was this changed? Seems a leftover change from your landscape orientation PR.  
Please provide context, or remove.

taylorpatterson-T1D replied:
This added the space needed to prevent wrapping in the legend.

Sjoerd-Bo3 replied:
Is this only on your device? did you make sure this is not weird now on other size devices?

taylorpatterson-T1D replied:
Tested in the iPhone SE simulator with a setting of 75 (to be sure) and it appears good, so I manually edited this file accordingly.
<img width="365" height="737" alt="Screenshot 2025-08-19 at 2 04 17 PM" src="https://github.com/user-attachments/assets/bdb657a7-06a5-4ac0-902d-ff7831e9aeb5" />

dnzxy commented on Trio/Sources/Modules/Treatments/View/TreatmentsRootView.swift:
```diff
@@ -86,6 +86,7 @@ extension Treatments {
             HStack {
                 HStack {
                     Text("Protein")
+                        .fixedSize(horizontal: true, vertical: false) // Prevent wrapping
```
What wrapping?

taylorpatterson-T1D replied:
"Pro 
tein"

was wrapping in the legend of that screen because the columns weren't spaced apart.

Sjoerd-Bo3 replied:
Is this only on your device? did you make sure this is not weird now on other size devices?

dnzxy commented on Trio/Sources/Router/Screen.swift:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/Services/ActivityDetectionManager.swift:
Unnecessary changes, plus introduces a new feature on top of a proposed bugfix.
Please remove / revert.

dnzxy commented on Trio/Sources/Services/AppVersionChecker/AppVersionChecker.swift:
Unnecessary linting change, please revert.

dnzxy reviewed (changes requested):
Hi Taylor @taylorpatterson-T1D ,
as Sjoerd wrote, this is correctly applied now, but mixes a proposed bugfix with a proposed feature. I've commented on all files that need to be removed / reverted and added some questions around changes you are proposing.

taylorpatterson-T1D commented:
@dnzxy @Sjoerd-Bo3 Thank you both for your feedback. I appreciate you taking the time. I do not know what I am doing with pushing changes using the tools I have and that is screwing me up. I apologize. Originally I had trouble just switching to the dev branch of Trio, which is where I wanted to make changes. Kept getting a submodule error on Danakit and could not resolve it using Github Desktop (screenshot). 

<img width="1146" height="809" alt="Error switching to dev branch of Trio" src="https://github.com/user-attachments/assets/0f9f95de-4598-4978-be6b-fd69f0144893" />

So instead I followed the process to rebase and push just the 4 files that were part of the bug fix first. I do not know why or how the other 17 files from my AutoApplyOverride feature were included.

I am still learning with the devops side of the equation. I am not a professional developer, just the idea man, armed with an AI coder trying to contain changes to new files as much as possible so I change very few things about the existing files of Trio.

I will close this PR, go back to the drawing board, and try again. I will ONLY submit the 4 file bug fix change first, then wait for your confirmation before moving on to the feature.

Sjoerd-Bo3 commented:
@taylorpatterson-T1D Tip: dont use github desktop :) 
If using mac, just use the VS Code integrated GIT client. It will work for now. Or Try GIT Tower (which I use and Love)

Just create a new branch from DEV and commit to there. And push it to you remote and PR it. There are also free basic GIT course that will help you with these basic things.

Also with using an AI, Please just do step by step. Not like 10 things at a time. If you send me the requirements, I can think with you on this aswell.

But to be honest, AI is nice. But it also creates a lot of crap if you are not knowing what you are doing and what it produces. And can review it yourself.

taylorpatterson-T1D committed:
Delete Trio.xcodeproj/project.pbxproj

taylorpatterson-T1D committed:
Delete Trio/Sources/APS/APSManager.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Assemblies/ServiceAssembly.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Assemblies/UIAssembly.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Models/TrioSettings.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Modules/AutoApplyOverride/AutoApplyOverrideDataFlow.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Modules/AutoApplyOverride/AutoApplyOverrideProvider.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Modules/AutoApplyOverride/AutoApplyOverrideStateModel.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Modules/AutoApplyOverride/View/AutoApplyOverrideRootView.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Modules/Settings/View/SettingsRootView.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Modules/Settings/View/Subviews/FeatureSettingsView.swift

taylorpatterson-T1D committed:
Update MealStatsView.swift

adjusted spacer from 90 to 75 to be safe for smaller devices. Tested ok in iPhone SE

taylorpatterson-T1D committed:
Delete Trio/Sources/Router/Screen.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Services/ActivityDetectionManager.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Services/AppVersionChecker/AppVersionChecker.swift

taylorpatterson-T1D committed:
Delete Trio/Sources/Localizations/Main/Localizable.xcstrings

taylorpatterson-T1D committed:
Delete Trio/Resources/Info.plist

taylorpatterson-T1D committed:
Delete Trio/Resources/InfoPlist.xcstrings

taylorpatterson-T1D commented:
I believe I removed the files related to the AutoApplyOverride changes. My apologies. The only 4 files related to the bug fixes are:
[Trio/Sources/Modules/Home/View/Chart/ChartElements/SelectionPopoverView.swift](https://github.com/nightscout/Trio/pull/753/files#diff-64e0e21a46c1619b7ebf1602ea8770d0a83c5b0fee8ae0420ec3eea719500071)
[Trio/Sources/Modules/Home/View/Chart/MainChartView.swift](https://github.com/nightscout/Trio/pull/753/files#diff-b31a9272b7990c5bb3e2c14937ceb5549c80df7fec71e09b4cbef6b9f1f9edc2)
[Trio/Sources/Modules/Stat/View/ViewElements/Meal/MealStatsView.swift](https://github.com/nightscout/Trio/pull/753/files#diff-94d9c47f37681858327236e7e3513262d0a56d129975c4e1421114e7cff20f00)
[Trio/Sources/Modules/Treatments/View/TreatmentsRootView.swift](https://github.com/nightscout/Trio/pull/753/files#diff-4d5df633b4d7be4179375e3fbfbfd291939952089f51c06328397f317491183c)

dnzxy commented:
@taylorpatterson-T1D a quick look at the file diff tells me that you *overall deleted* a bunch of files from the project, not just from your PR. What we "need" here is not so much removing files from the project, but not to include any changes other than the ones to those 4 files you have linked, and then have that as part of your PR. Whatever you have attempted to do here, something went wrong at some point. These deletions are not correct.
<img width="786" height="785" alt="image" src="https://github.com/user-attachments/assets/c87fcd12-964b-40cf-8abf-91fb80051815" />

taylorpatterson-T1D commented:
@dnzxy I deleted them from this PR (#753) because I wanted to match it's description to what we're actually changing only (the remaining 4 files), as I thought you were guiding me to do. Then I planned to submit a separate PR for the AutoApplyOverride feature later and as it's own feature branch. I have a 3rd PR coming for a new feature FoodFinder for Trio but I'm not even going to go there until I learn enough to not be so dangerous.

Honestly, I wish someone could just take my files, make sure they comply with Trio's architecture, then post the PR correctly on my behalf. I don't need credit and I'd rather see people use the new features and not have to wait for me to learn this stuff. I fear the value of the features will decay. :-)

dnzxy commented:
I asked you to remove it from the PR, not delete it from the project. 

Sorry, I / we cannot teach you how to use git. Use GPT or Claude and a terminal of choice and let it guide you to do the things you want to do.

Don‘t take this the wrong way please, contributions are always welcome and no one is born a developer or a git expert — heck, even its creator called it "**g**oddamn **i**diotic **t**ruckload of shit" when it breaks. Yet, I‘m somewhat concerned and am calling the quality of your features and how you (?) implement(ed) them very much into question if you are not able to do the bare minimum git. It’s part of this game we play here.

dnzxy commented:
Closing this PR as it's not remotely in a state to me merged and there has been no progress here in ~11 weeks.
